File: icarus/src/sensors.rs
//
// sensors.rs
//
// @author Natesh Narain <nnaraindev@gmail.com>
// @date Dec 01 2021
//

pub mod imu {
    use mpu6050::Mpu6050;
    use crate::types::I2cBus;

    // Re-exports
    pub type Imu<'a> = Mpu6050<I2cBus<'a>>;

    pub fn init(i2c: I2cBus) -> Imu {
        Mpu6050::new_with_addr(i2c, 0x68)
    }
}

pub mod barometer {
    use bmp388::BMP388;
    use crate::types::I2cBus;
    use crate::IcarusError;

    pub type Barometer<'a> = BMP388<I2cBus<'a>>;

    pub fn init(i2c: I2cBus) -> Result<Barometer, IcarusError> {
        BMP388::new_with_addr(i2c, 0x77).map_err(|_| IcarusError::HardwareInitError)
    }
}

File: lib/sensors/Sensors.cpp
//
// Sensors.cpp
//
// @author Natesh Narain <nnaraindev@gmail.com>
// @date Feb 03 2023
//

#include <Arduino.h>
#include "Sensors.hpp"

#include <Wire.h>

// IMU Address
#define MPU_ADDRESS 0x68
// Magnetometer Address
#define BMM_ADDRESS 0x10
// Barometer Address
#define BMP_ADDRESS 0x77

bool Sensors::begin(uint8_t scl, uint8_t sda)
{
    const auto wire_ok = Wire.setPins(sda, scl) && Wire.begin();
    // Wire.setClock()

    if (!wire_ok)
    {
        Serial.println("Failed to initialize Wire protocol");
    }

    mpu.setDeviceAddress(MPU_ADDRESS);

    const auto mpu_ok = mpu.begin();

    if (!mpu_ok)
    {
        Serial.println("Failed to initialize MPU");
    }
    else
    {
        mpu.calibrate();
    }

    return wire_ok && mpu_ok;
}

void Sensors::update()
{
    mpu.update();
}

attitude_t Sensors::getAttitude()
{
    return mpu.getAttitude();
}